README.md
[![Build Status](https://travis-ci.org/broadinstitute/cromwell.svg?branch=develop)](https://travis-ci.org/broadinstitute/cromwell?branch=develop)
[![Coverage Status](https://coveralls.io/repos/broadinstitute/cromwell/badge.svg?branch=develop)](https://coveralls.io/r/broadinstitute/cromwell?branch=develop)
[![Join the chat at https://gitter.im/broadinstitute/cromwell](https://badges.gitter.im/broadinstitute/cromwell.svg)](https://gitter.im/broadinstitute/cromwell?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

Cromwell
========

A [Workflow Management System](https://en.wikipedia.org/wiki/Workflow_management_system) geared towards [scientific workflows](https://en.wikipedia.org/wiki/Scientific_workflow_system). Cromwell is open sourced under the BSD 3-Clause license.

<!---toc start-->

* [Getting Help](#getting-help)
  * [Website and User Guide](#website-and-user-guide)
  * [Support Forum](#support-forum)
  * [Gitter](#gitter)
* [Requirements](#requirements)
* [Building](#building)
* [Installing](#installing)
  * [Upgrading from 0.19 to 0.21](#upgrading-from-019-to-021)
* [Command Line Usage](#command-line-usage)
* [Getting Started with WDL](#getting-started-with-wdl)
  * [WDL Support](#wdl-support)
* [Configuring Cromwell](#configuring-cromwell)
  * [Workflow Submission](#workflow-submission)
  * [Database](#database)
  * [SIGINT abort handler](#sigint-abort-handler)
* [Security](#security)
* [Backends](#backends)
  * [Backend Filesystems](#backend-filesystems)
    * [Shared Local Filesystem](#shared-local-filesystem)
    * [Google Cloud Storage Filesystem](#google-cloud-storage-filesystem)
  * [Local Backend](#local-backend)
  * [Google JES Backend](#google-jes-backend)
    * [Configuring Google Project](#configuring-google-project)
    * [Configuring Authentication](#configuring-authentication)
      * [Application Default Credentials](#application-default-credentials)
      * [Service Account](#service-account)
      * [Refresh Token](#refresh-token)
    * [Docker](#docker)
    * [Monitoring](#monitoring)
  * [GA4GH TES Backend](#ga4gh-tes-backend)
    * [Configuring](#configuring)
    * [Supported File Systems](#supported-file-systems)
    * [Docker](#docker)
    * [CPU, Memory and Disk](#cpu-memory-and-disk)
  * [Sun GridEngine Backend](#sun-gridengine-backend)
  * [HtCondor Backend](#htcondor-backend)
    * [Caching configuration](#caching-configuration)
    * [Docker](#docker)
    * [CPU, Memory and Disk](#cpu-memory-and-disk)
    * [Native Specifications](#native-specifications)
  * [Spark Backend](#spark-backend)
    * [Configuring Spark Project](#configuring-spark-project)
    * [Configuring Spark Master and Deploy Mode](#configuring-spark-master-and-deploy-mode)
    * [Spark runtime attributes](#spark-runtime-attributes)
    * [Spark Environment](#spark-environment)
    * [Sample Wdl](#sample-wdl)
* [Runtime Attributes](#runtime-attributes)
  * [Specifying Default Values](#specifying-default-values)
  * [continueOnReturnCode](#continueonreturncode)
  * [cpu](#cpu)
  * [disks](#disks)
    * [Boot Disk](#boot-disk)
  * [zones](#zones)
  * [docker](#docker)
  * [failOnStderr](#failonstderr)
  * [memory](#memory)
  * [preemptible](#preemptible)
* [Logging](#logging)
* [Workflow Options](#workflow-options)
* [Labels](#labels)
  * [Custom Labels File](#custom-labels-file)
  * [Label Format](#label-format)
* [Call Caching](#call-caching)
  * [Configuring Call Caching](#configuring-call-caching)
  * [Call Caching Workflow Options](#call-caching-workflow-options)
  * [Local Filesystem Options](#local-filesystem-options)
* [Imports](#imports)
* [Sub Workflows](#sub-workflows)
  * [Execution](#execution)
  * [Metadata](#metadata)
* [REST API](#rest-api)
  * [REST API Versions](#rest-api-versions)
  * [POST /api/workflows/:version](#post-apiworkflowsversion)
  * [POST /api/workflows/:version/batch](#post-apiworkflowsversionbatch)
  * [GET /api/workflows/:version/query](#get-apiworkflowsversionquery)
  * [POST /api/workflows/:version/query](#post-apiworkflowsversionquery)
  * [GET /api/workflows/:version/:id/status](#get-apiworkflowsversionidstatus)
  * [GET /api/workflows/:version/:id/outputs](#get-apiworkflowsversionidoutputs)
  * [GET /api/workflows/:version/:id/timing](#get-apiworkflowsversionidtiming)
  * [GET /api/workflows/:version/:id/logs](#get-apiworkflowsversionidlogs)
  * [GET /api/workflows/:version/:id/metadata](#get-apiworkflowsversionidmetadata)
  * [POST /api/workflows/:version/:id/abort](#post-apiworkflowsversionidabort)
  * [GET /api/workflows/:version/backends](#get-apiworkflowsversionbackends)
  * [GET /api/workflows/:version/callcaching/diff](#get-apiworkflowsversioncallcachingdiff)
  * [GET /api/engine/:version/stats](#get-apiengineversionstats)
  * [GET /api/engine/:version/version](#get-apiengineversionversion)
  * [Error handling](#error-handling)
* [Developer](#developer)
  * [Generating table of contents on Markdown files](#generating-table-of-contents-on-markdown-files)
  * [Generating and Hosting ScalaDoc](#generating-and-hosting-scaladoc)

<!---toc end-->

# Getting Help

## Website and User Guide

The [WDL website](https://software.broadinstitute.org/wdl/) is the best place to go for more information on both WDL and Cromwell. In particular new users should check out the [user guide](https://software.broadinstitute.org/wdl/userguide/) which has many tutorials, examples and other bits to get you started.

## Support Forum

If you have questions that aren't covered by the website you can ask them in the [support forum](http://gatkforums.broadinstitute.org/wdl/categories/ask-the-wdl-team).

## Gitter
There is a [Cromwell gitter channel](https://gitter.im/broadinstitute/cromwell) where people can discuss Cromwell and related topics with both the developers and user community.

# Requirements

The following is the toolchain used for development of Cromwell.  Other versions may work, but these are recommended.

* [Scala 2.11.8](http://www.scala-lang.org/news/2.11.8/)
* [SBT 0.13.12](https://github.com/sbt/sbt/releases/tag/v0.13.12)
* [Java 8](http://www.oracle.com/technetwork/java/javase/overview/java8-2100321.html)

# Building

`sbt assembly` will build a runnable JAR in `target/scala-2.11/`

Tests are run via `sbt test`.  Note that the tests do require Docker to be running.  To test this out while downloading the Ubuntu image that is required for tests, run `docker pull ubuntu:latest` prior to running `sbt test`

# Installing

OS X users can install Cromwell with Homebrew: `brew install cromwell`.

## Upgrading from 0.19 to 0.21

See the [migration document](MIGRATION.md) for more details.

# Command Line Usage

For built-in documentation of Cromwell command line usage, run the Cromwell JAR file with no arguments:

```
$ java -jar cromwell-<versionNumber>.jar
```

For example, `$ java -jar cromwell-28.jar`. You will get a usage message like the following:

```
cromwell 28-78824f3-SNAP
Usage: java -jar /path/to/cromwell.jar [server|run] [options] <args>...

  --help                   Cromwell - Workflow Execution Engine
  --version                
Command: server
Starts a web server on port 8000.  See the web server documentation for more details about the API endpoints.
Command: run [options] workflow-source
Run the workflow and print out the outputs in JSON format.
  workflow-source          Workflow source file.
  -i, --inputs <value>     Workflow inputs file.
  -o, --options <value>    Workflow options file.
  -t, --type <value>       Workflow type.
  -v, --type-version <value>
                           Workflow type version.
  -l, --labels <value>     Workflow labels file.
  -p, --imports <value>    A directory or zipfile to search for workflow imports.
  -m, --metadata-output <value>
                           An optional directory path to output metadata.
```

## --version

The `--version` option prints the version of Cromwell and exits.

## --help

The `--help` option prints the full help text above and exits.

## server

The `server` command runs Cromwell as a web server.  No arguments are accepted.
See the documentation for Cromwell's REST endpoints [here](#rest-api).

## run

The `run` command executes a single workflow in Cromwell.

### workflow-source
The `run` command requires a single argument for the workflow source file.
 
### --inputs
Workflow inputs are technically optional, but highly recommended so that inputs are not hardcoded into the workflow
source file.

### --options
An optional file of workflow options.  Some options are global (supported by all backends), while others are backend-specific.

### --type
An optional parameter to specify the language for the workflow source.  Any value specified for this parameter is currently
ignored and internally the value `WDL` is used.

### --type-version
An optional parameter to specify the version of the language for the workflow source.  Currently any specified value is ignored.

### --labels
An optional parameter to specify a file of JSON key-value label pairs to associate with the workflow.

### --imports
You have the option of importing WDL workflows or tasks to use within your workflow, known as sub-workflows.
If you use sub-workflows within your primary workflow then you must include a zip file with the WDL import files.

For example, say you have a directory of WDL files:

```
wdl_library
└──cgrep.wdl
└──ps.wdl
└──wc.wdl
```

If you zip that directory into `wdl_library.zip`, then you can reference and use these WDLs within your primary WDL.

This could be your primary WDL:

```
import "ps.wdl" as ps
import "cgrep.wdl"
import "wc.wdl" as wordCount

workflow my_wf {

call ps.ps as getStatus
call cgrep.cgrep { input: str = getStatus.x }
call wordCount { input: str = ... }

}
```

Then to run this WDL without any inputs, workflow options, or metadata files, you would enter:

`$ java -jar cromwell-<versionNumber>.jar run my_wf.wdl --imports /path/to/wdl_library.zip`

### --metadata-output

You can include a path where Cromwell will write the workflow metadata JSON, such as start/end timestamps, status, inputs, and outputs. By default, Cromwell does not write workflow metadata.

This example includes a metadata path called `/path/to/my_wf.metadata`:

```
$ java -jar cromwell-<versionNumber>.jar run my_wf.wdl --metadata-output /path/to/my_wf.metadata
```

Again, Cromwell is very verbose. Here is the metadata output in my_wf.metadata:

```
{
  "workflowName": "my_wf",
  "submittedFiles": {
    "inputs": "{\"my_wf.hello.addressee\":\"m'Lord\"}",
    "workflow": "\ntask hello {\n  String addressee\n  command {\n    echo \"Hello ${addressee}!\"\n  }\n  output {\n    String salutation = read_string(stdout())\n  }\n  runtime {\n   
\n  }\n}\n\nworkflow my_wf {\n  call hello\n  output {\n     hello.salutation\n  }\n}\n",
    "options": "{\n\n}"
  },
  "calls": {
    "my_wf.hello": [
      {
        "executionStatus": "Done",
        "stdout": "/Users/jdoe/Documents/cromwell-executions/my_wf/cd0fe94a-984e-4a19-ab4c-8f7f07038068/call-hello/execution/stdout",
        "backendStatus": "Done",
        "shardIndex": -1,
        "outputs": {
          "salutation": "Hello m'Lord!"
        },
        "runtimeAttributes": {
          "continueOnReturnCode": "0",
          "failOnStderr": "false"
        },
        "callCaching": {
          "allowResultReuse": false,
          "effectiveCallCachingMode": "CallCachingOff"
        },
        "inputs": {
          "addressee": "m'Lord"
        },
        "returnCode": 0,
        "jobId": "28955",
        "backend": "Local",
        "end": "2017-04-19T10:53:25.045-04:00",
        "stderr": "/Users/jdoe/Documents/cromwell-executions/my_wf/cd0fe94a-984e-4a19-ab4c-8f7f07038068/call-hello/execution/stderr",
        "callRoot": "/Users/jdoe/Documents/cromwell-executions/my_wf/cd0fe94a-984e-4a19-ab4c-8f7f07038068/call-hello",
        "attempt": 1,
        "executionEvents": [
          {
            "startTime": "2017-04-19T10:53:23.570-04:00",
            "description": "PreparingJob",
            "endTime": "2017-04-19T10:53:23.573-04:00"
          },
          {
            "startTime": "2017-04-19T10:53:23.569-04:00",
            "description": "Pending",
            "endTime": "2017-04-19T10:53:23.570-04:00"
          },
          {
            "startTime": "2017-04-19T10:53:25.040-04:00",
            "description": "UpdatingJobStore",
            "endTime": "2017-04-19T10:53:25.045-04:00"
          },
          {
            "startTime": "2017-04-19T10:53:23.570-04:00",
            "description": "RequestingExecutionToken",
            "endTime": "2017-04-19T10:53:23.570-04:00"
          },
          {
            "startTime": "2017-04-19T10:53:23.573-04:00",
            "description": "RunningJob",
            "endTime": "2017-04-19T10:53:25.040-04:00"
          }
        ],
        "start": "2017-04-19T10:53:23.569-04:00"
      }
    ]
  },
  "outputs": {
    "my_wf.hello.salutation": "Hello m'Lord!"
  },
  "workflowRoot": "/Users/jdoe/Documents/cromwell-executions/my_wf/cd0fe94a-984e-4a19-ab4c-8f7f07038068",
  "id": "cd0fe94a-984e-4a19-ab4c-8f7f07038068",
  "inputs": {
    "my_wf.hello.addressee": "m'Lord"
  },
  "submission": "2017-04-19T10:53:19.565-04:00",
  "status": "Succeeded",
  "end": "2017-04-19T10:53:25.063-04:00",
  "start": "2017-04-19T10:53:23.535-04:00"
}
```

# Getting Started with WDL

For many examples on how to use WDL see [the WDL site](https://github.com/broadinstitute/wdl#getting-started-with-wdl)

## WDL Support

:pig2: Cromwell supports the following subset of WDL features:

* [Language Specification](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#language-specification)
  * [Whitespace, Strings, Identifiers, Constants](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#whitespace-strings-identifiers-constants)
  * [Types](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#types)
  * [Fully Qualified Names & Namespaced Identifiers](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#fully-qualified-names--namespaced-identifiers)
  * [Declarations](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#declarations)
  * [Expressions](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#expressions)
  * [Operator Precedence Table](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#operator-precedence-table)
  * [Member Access](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#member-access)
  * [Map and Array Indexing](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#map-and-array-indexing)
  * [Function Calls](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#function-calls)
  * [Array Literals](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#array-literals)
  * [Map Literals](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#map-literals)
* [Document](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#document)
* [Task Definition](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#task-definition)
  * [Sections](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#sections)
  * [Command Section](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#command-section)
    * [Command Parts](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#command-parts)
    * [Command Part Options](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#command-part-options)
      * [sep](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#sep)
      * [true and false](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#true-and-false)
      * [default](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#default)
    * [Alternative heredoc syntax](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#alternative-heredoc-syntax)
    * [Stripping Leading Whitespace](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#stripping-leading-whitespace)
  * [Outputs Section](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#outputs-section)
  * [String Interpolation](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#string-interpolation)
  * [Runtime Section](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#runtime-section)
    * [docker](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#docker)
    * [memory](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#memory)
  * [Parameter Metadata Section](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#parameter-metadata-section)
  * [Metadata Section](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#metadata-section)
* [Workflow Definition](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#workflow-definition)
  * [Call Statement](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#call-statement)
    * [Sub Workflows](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#sub-workflows)
  * [Scatter](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#scatter)
  * [Parameter Metadata](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#parameter-metadata)
  * [Metadata](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#metadata)
  * [Outputs](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#outputs)
* [Namespaces](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#namespaces)
* [Scope](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#scope)
* [Optional Parameters & Type Constraints](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#optional-parameters--type-constraints)
  * [Prepending a String to an Optional Parameter](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#prepending-a-string-to-an-optional-parameter)
* [Scatter / Gather](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#scatter--gather)
* [Variable Resolution](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#variable-resolution)
  * [Task-Level Resolution](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#task-level-resolution)
  * [Workflow-Level Resolution](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#workflow-level-resolution)
* [Computing Inputs](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#computing-inputs)
  * [Task Inputs](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#task-inputs)
  * [Workflow Inputs](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#workflow-inputs)
  * [Specifying Workflow Inputs in JSON](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#specifying-workflow-inputs-in-json)
* [Type Coercion](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#type-coercion)
* [Standard Library](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#standard-library)
  * [File stdout()](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#file-stdout)
  * [File stderr()](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#file-stderr)
  * [Array\[String\] read_lines(String|File)](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#arraystring-read_linesstringfile)
    * File reads are limited to 128 KB. Configurable via conf file.
  * [Array\[Array\[String\]\] read_tsv(String|File)](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#arrayarraystring-read_tsvstringfile)
    * File reads are limited to 128 KB. Configurable via conf file.
  * [Map\[String, String\] read_map(String|File)](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#mapstring-string-read_mapstringfile)
    * File reads are limited to 128 KB. Configurable via conf file.
  * [Object read_object(String|File)](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#object-read_objectstringfile)
    * File reads are limited to 128 KB. Configurable via conf file.
  * [Array\[Object\] read_objects(String|File)](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#arrayobject-read_objectsstringfile)
    * File reads are limited to 128 KB. Configurable via conf file.
  * [Int read_int(String|File)](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#int-read_intstringfile)
    * File reads are limited to 19 B. Configurable via conf file.
  * [String read_string(String|File)](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#string-read_stringstringfile)
    * File reads are limited to 128 KB. Configurable via conf file.
  * [Float read_float(String|File)](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#float-read_floatstringfile)
    * File reads are limited to 50 B. Configurable via conf file.
  * [Boolean read_boolean(String|File)](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#boolean-read_booleanstringfile)
    * File reads are limited to 7 B. Configurable via conf file.
  * [File write_lines(Array\[String\])](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#file-write_linesarraystring)
  * [File write_tsv(Array\[Array\[String\]\])](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#file-write_tsvarrayarraystring)
  * [File write_map(Map\[String, String\])](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#file-write_mapmapstring-string)
  * [File write_object(Object)](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#file-write_objectobject)
  * [File write_objects(Array\[Object\])](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#file-write_objectsarrayobject)
  * [Float size(File, \[String\])](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#float-sizefile-string)
  * [String sub(String, String, String)](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#string-substring-string-string)
  * [Array\[Int\] range(Int)](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#arrayint-rangeint)
  * [Array\[Array\[X\]\] transpose(Array\[Array\[X\]\])](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#arrayarrayx-transposearrayarrayx)
  * [Array\[Pair\[X,Y\]\] zip(Array\[X\], Array\[Y\])](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#arraypairxy-ziparrayx-arrayy)
  * [Array\[Pair\[X,Y\]\] cross(Array\[X\], Array\[Y\])](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#arraypairxy-crossarrayx-arrayy)
  * [Integer length(Array\[X\])](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#integer-lengtharrayx)
  * [Array\[String\] prefix(String, Array\[X\])](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#arraystring-prefixstring-arrayx)
* [Data Types & Serialization](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#data-types--serialization)
  * [Serialization of Task Inputs](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#serialization-of-task-inputs)
    * [Primitive Types](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#primitive-types)
    * [Compound Types](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#compound-types)
      * [Array serialization](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#array-serialization)
        * [Array serialization by expansion](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#array-serialization-by-expansion)
        * [Array serialization using write_lines()](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#array-serialization-using-write_lines)
      * [Map serialization](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#map-serialization)
        * [Map serialization using write_map()](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#map-serialization-using-write_map)
      * [Object serialization](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#object-serialization)
        * [Object serialization using write_object()](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#object-serialization-using-write_object)
      * [Array\[Object\] serialization](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#arrayobject-serialization)

CHANGELOG.md
# Cromwell Change Log

## 28

### Command Line Usage

Command line usage has been extensively revised for Cromwell 28.  Please see the
[README](https://github.com/broadinstitute/cromwell#command-line-usage) for details.

### Call Caching

* Hash values calculated by Cromwell for a call when call caching is enabled are now published to the metadata.
It is published even if the call failed. However if the call is attempted multiple times (because it has been preempted for example),
since hash values are strictly identical for all attempts, they will only be published in the last attempt section of the metadata for this call.
If the hashes fail to be calculated, the reason is indicated in a `hashFailures` field in the `callCaching` section of the call metadata.

See the [README](https://github.com/broadinstitute/cromwell#get-apiworkflowsversionidmetadata) for an example metadata response.

* New endpoint returning the hash differential for 2 calls. 

`GET /api/workflows/:version/callcaching/diff`

See the [README](https://github.com/broadinstitute/cromwell#get-apiworkflowsversioncallcachingdiff) for more details.

### Workflow Submission

The workflow submission parameters `wdlSource` and `wdlDependencies` have been deprecated in favor of `workflowSource` and
`workflowDependencies` respectively.  The older names are still supported in Cromwell 28 with deprecation warnings but will
be removed in a future version of Cromwell.

### JES Backend

The JES backend now supports a `filesystems.gcs.caching.duplication-strategy` configuration entry.
It can be set to specify the desired behavior of Cromwell regarding call outputs when a call finds a hit in the cache.
The default value is `copy` which will copy all output files to the new call directory.
A second value is allowed, `reference`, that will instead point to the original output files, without copying them.

```hocon
filesystems {
  gcs {
    auth = "application-default"
    
    caching {
      duplication-strategy = "reference"
    }
  }
}
```

A placeholder file will be placed in the execution folder of the cached call to explain the absence of output files and point to the location of the original ones.

## 27

### Migration

* Call Caching has been improved in this version of Cromwell, specifically the time needed to determine whether or not a job can be cached
 has drastically decreased. To achieve that the database schema has been modified and a migration is required in order to preserve the pre-existing cached jobs.
 This migration is relatively fast compared to previous migrations. To get an idea of the time needed, look at the size of your `CALL_CACHING_HASH_ENTRY` table.
 As a benchmark, it takes 1 minute for a table with 6 million rows.
 The migration will only be executed on MySQL. Other databases will lose their previous cached jobs.
 In order to run properly on MySQL, **the following flag needs to be adjusted**: https://dev.mysql.com/doc/refman/5.5/en/server-system-variables.html#sysvar_group_concat_max_len
 The following query will give you a minimum to set the group_concat_max_len value to:
 
 ```sql
SELECT MAX(aggregated) as group_concat_max_len FROM
      (
            SELECT cche.CALL_CACHING_ENTRY_ID, SUM(LENGTH(CONCAT(cche.HASH_KEY, cche.HASH_VALUE))) AS aggregated
            FROM CALL_CACHING_HASH_ENTRY cche
            GROUP BY cche.CALL_CACHING_ENTRY_ID
      ) aggregation
 ```

 Here is the SQL command to run to set the group_concat_max_len flag to the proper value:
 
 ```sql
SET GLOBAL group_concat_max_len = value
 ```
 
 Where `value` is replaced with the value you want to set it to.
 
 Note that the migration will fail if the flag is not set properly.
 
### Breaking Changes

* The update to Slick 3.2 requires a database stanza to
[switch](http://slick.lightbend.com/doc/3.2.0/upgrade.html#profiles-vs-drivers) from using `driver` to `profile`.

```hocon
database {
  #driver = "slick.driver.MySQLDriver$" #old
  profile = "slick.jdbc.MySQLProfile$"  #new
  db {
    driver = "com.mysql.jdbc.Driver"
    url = "jdbc:mysql://host/cromwell?rewriteBatchedStatements=true"
    user = "user"
    password = "pass"
    connectionTimeout = 5000
  }
}
```

### Call Caching

Cromwell now supports call caching with floating Docker tags (e.g. `docker: "ubuntu:latest"`). Note it is still considered
a best practice to specify Docker images as hashes where possible, especially for production usages.

Within a single workflow Cromwell will attempt to resolve all floating tags to the same Docker hash, even if Cromwell is restarted
during the execution of a workflow. In call metadata the `docker` runtime attribute is now the same as the
value that actually appeared in the WDL:

```
   "runtimeAttributes": {
     "docker": "ubuntu:latest",
     "failOnStderr": "false",
     "continueOnReturnCode": "0"
   }
```

Previous versions of Cromwell rewrote the `docker` value to the hash of the Docker image.

There is a new call-level metadata value `dockerImageUsed` which captures the hash of the Docker image actually used to
run the call:

```
   "dockerImageUsed": "library/ubuntu@sha256:382452f82a8bbd34443b2c727650af46aced0f94a44463c62a9848133ecb1aa8"
```

### Docker

* The Docker section of the configuration has been slightly reworked 
An option to specify how a Docker hash should be looked up has been added. Two methods are available.
    "local" will try to look for the image on the machine where cromwell is running. If it can't be found, Cromwell will try to `pull` the image and use the hash from the retrieved image.
    "remote" will try to look up the image hash directly on the remote repository where the image is located (Docker Hub and GCR are supported)
Note that the "local" option will require docker to be installed on the machine running cromwell, in order for it to call the docker CLI.
* Adds hash lookup support for public [quay.io](https://quay.io/) images.

### WDL Feature Support

src/main/scala/cromwell/CommandLineParser.scala
package cromwell

import com.typesafe.config.ConfigFactory
import cromwell.core.path.{DefaultPathBuilder, Path}
import scopt.OptionParser

object CommandLineParser extends App {

  sealed trait Command
  case object Run extends Command
  case object Server extends Command

  case class CommandLineArguments(command: Option[Command] = None,
                                  workflowSource: Option[Path] = None,
                                  workflowInputs: Option[Path] = None,
                                  workflowOptions: Option[Path] = None,
                                  workflowType: Option[String] = Option("WDL"),
                                  workflowTypeVersion: Option[String] = Option("v2.0-draft"),
                                  workflowLabels: Option[Path] = None,
                                  imports: Option[Path] = None,
                                  metadataOutput: Option[Path] = None
                                 )

  lazy val cromwellVersion = ConfigFactory.load("cromwell-version.conf").getConfig("version").getString("cromwell")

  case class ParserAndCommand(parser: OptionParser[CommandLineArguments], command: Option[Command])

//   cromwell 28-6ace4bc-SNAP
//
//   Usage: java -jar /path/to/cromwell.jar [server|run] [options] <args>...
//
//
//     --help                   Cromwell - Workflow Execution Engine
//     --version
//   Command: server
//   Starts a web server on port 8000.  See the web server documentation for more details about the API endpoints.
//   Command: run [options] workflow-source
//   Run the workflow and print out the outputs in JSON format.
//     workflow-source          Workflow source file.
//     -i, --inputs <value>     Workflow inputs file.
//     -o, --options <value>    Workflow options file.
//     -t, --type <value>       Workflow type.
//     -v, --type-version <value>
//                              Workflow type version.
//     -l, --labels <value>     Workflow labels file.
//     -p, --imports <value>    A directory or zipfile to search for workflow imports.
//     -m, --metadata-output <value>
//                              An optional directory path to output metadata.

  def buildParser(): scopt.OptionParser[CommandLineArguments] = {
    new scopt.OptionParser[CommandLineArguments]("java -jar /path/to/cromwell.jar") {
      head("cromwell", cromwellVersion)

      help("help").text("Cromwell - Workflow Execution Engine")

      version("version")

      cmd("server").action((_, c) => c.copy(command = Option(Server))).text(
        "Starts a web server on port 8000.  See the web server documentation for more details about the API endpoints.")

      cmd("run").
        action((_, c) => c.copy(command = Option(Run))).
        text("Run the workflow and print out the outputs in JSON format.").
        children(
          arg[String]("workflow-source").text("Workflow source file.").required().
            action((s, c) => c.copy(workflowSource = Option(DefaultPathBuilder.get(s)))),
          opt[String]('i', "inputs").text("Workflow inputs file.").
            action((s, c) =>
              c.copy(workflowInputs = Option(DefaultPathBuilder.get(s)))),
          opt[String]('o', "options").text("Workflow options file.").
            action((s, c) =>
              c.copy(workflowOptions = Option(DefaultPathBuilder.get(s)))),
          opt[String]('t', "type").text("Workflow type.").
            action((s, c) =>
              c.copy(workflowType = Option(s))),
          opt[String]('v', "type-version").text("Workflow type version.").
            action((s, c) =>
              c.copy(workflowTypeVersion = Option(s))),
          opt[String]('l', "labels").text("Workflow labels file.").
            action((s, c) =>
              c.copy(workflowLabels = Option(DefaultPathBuilder.get(s)))),
          opt[String]('p', "imports").text(
            "A directory or zipfile to search for workflow imports.").
            action((s, c) =>
              c.copy(imports = Option(DefaultPathBuilder.get(s)))),
          opt[String]('m', "metadata-output").text(
            "An optional directory path to output metadata.").
            action((s, c) =>
              c.copy(metadataOutput = Option(DefaultPathBuilder.get(s))))
        )
    }
  }

  def runCromwell(args: CommandLineArguments): Unit = {
    args.command match {
      case Some(Run) => CromwellEntryPoint.runSingle(args)
      case Some(Server) => CromwellEntryPoint.runServer()
      case None => parser.showUsage()
    }
  }

  val parser = buildParser()

  val parsedArgs = parser.parse(args, CommandLineArguments())
  parsedArgs match {
    case Some(pa) => runCromwell(pa)
    case None => parser.showUsage()
  }
}
